compiler/src/scala/quoted/runtime/impl/QuoteMatcher.scala

```diff
@@ -465,14 +510,43 @@ object QuoteMatcher {
           case Block(List(DefDef(nme.ANON_FUN, _, _, Apply(Ident(name), _))), _) => name.asTermName
           case arg => arg.symbol.name.asTermName
         }
-        val paramTypes = args.map(x => mapTypeHoles(x.tpe.widenTermRefExpr))
+        val paramTypes = args.map(x => adaptTypes(mapTypeHoles(x.tpe.widenTermRefExpr)))
         val methTpe = MethodType(names)(_ => paramTypes, _ => mapTypeHoles(patternTpe))
         val meth = newAnonFun(ctx.owner, methTpe)
         def bodyFn(lambdaArgss: List[List[Tree]]): Tree = {
           val argsMap = args.view.map(_.symbol).zip(lambdaArgss.head).toMap
           val body = new TreeMap {
             override def transform(tree: Tree)(using Context): Tree =
               tree match
+                /*
+                 * When matching a method call `f(0)` against a HOAS pattern `p(g)` where
+                 * f has a method type `(x: Int): Int` and  `f` maps to `g`, `p` should be `g.apply(0)`
+                 * because the type of `g` is `Int => Int` due to eta expansion.
+                 *
+                 * Remaining TODOs from issue-17105
+                 * * [x] cover the case of nested method call
+                 * * [ ] contextual params?
+                 * * [ ] erasure types?
+                 * * [x] eta-expand only HOAS param methods
+                 */
+                case Apply(methId: Ident, args) =>
+                  env.get(tree.symbol).flatMap(argsMap.get)
+                    .map(fnId => ctx.typer.typed(
+                      untpd.Apply(
```

Comment:
Is this the right pattern to construct typed tree?

Reply:
Typing an untyped tree is dangerous. Depending on the context where this is typed, we could end up with different result. It is also a bit inefficient.

The correct way to do this is to create the typed tree directly. In this case you can use the `select` and `appliedToArgs` helper methods to construct this tree. The code should be `fnId.select(nme.apply).appliedToArgs(args))`.